File: ProjectEuler/p12.cc
//author agis daniels
//What is the value of the first triangle number to have over five hundred divisors?
#include <iostream> // cin cout
#include <math.h>   // sqrt
#include <vector>   // for the vector

using namespace std;

typedef vector<int> PrimesSet;

typedef vector<bool> PrimesBool;

typedef const int cstInt;

//get_div(const int x)
//get the number of divisors 
//param x is the number i am finding the divisors for
//return the amount of divisors
int get_div(cstInt x){
    int count=0;
    cstInt lim = sqrt(x);
    for(int i =1; i < lim ;i++){
        if(x%i==0){
            count++;
        }
    }
    count+=count;
    return count+1;
}

//get_sieve of eratosthenes(const int size)
//sieve of eratosthenes the bool part of it
//param size is the size we would like to cover
//return the seive in a bit set
PrimesBool get_sieve_of_eratosthenes(cstInt size, cstInt size2){
    cstInt r=(sqrt(size)-1)/2;
    PrimesBool seive(size2,0);
    seive[0]=1;
    for(int i=1; i<=r; ++i){
        if(0==seive[i]){
            for(int j=i*(i+1)+i*(i+1); j<size2; j+=(2*i+1)){
                seive[j]=1;
            }
        }
    }
    return seive;
}

//get_sieve(const int size)
//sieve of eratosthenes the vector part
//param size used for the size of the sieve
//return the seive in a vector
PrimesSet get_seive(cstInt size){
    cstInt size2=size/2; 
    PrimesBool seive=get_sieve_of_eratosthenes(size, size2);
    PrimesSet rVec;
    rVec.push_back(2);
    for(int i=0; i<size2; ++i){
        if(0==seive[i]){
            rVec.push_back(i+i+1);
        }
    }
    return rVec;
}

//get_ans_1(PrimesSet primeSeive)
//sieve of eratosthenes the vector part
//param sPrimesSet primeSeive a seive of primes to use on this problem
//return the seive in a vector
int get_ans_1(){
    int t=1, a=1, cnt=0, tt, expo;
    PrimesSet primeSeive=get_seive(65500); // get our main prime seive
    cstInt seiveSize=primeSeive.size();
    for(; cnt<=500; ){
        cnt=1;
        ++a;
        t+=a;
        tt=t;
        for(int i=0; i<seiveSize; ++i){
            cstInt curPrime=primeSeive[i];
            if(curPrime*curPrime>tt){
                cnt+=cnt;
                break;
            }
            expo=1;
            while(0==(tt%curPrime)){
                ++expo;
                tt/=curPrime;
            }
            if(1<expo){
                cnt*=expo;
            }if(1==tt){
                break;
            }
        }
    }
    return t;
}

//get_tri_num(cstInt num)
//returns the nth triangle number
int get_tri_num(int num){
    return (num*num-num)/2;
}

//get_ans_2(
//more faster version
//none params
//return the ans
int get_ans_2(){
    int n  =3; //starting prime
    int Dn =2; //num div for each prime
    int cnt=0; //to insure loop is entered
    int n1, Dn1, expo;
    PrimesSet primeSeive=get_seive(1000);
    cstInt seiveSize=primeSeive.size();
    for(; cnt<=500; ){
        ++n;
        n1=n;
        if(0==(n1%2)){
            n1/=2;   
        }
        Dn1=1;
        for(int i=0; i<seiveSize; ++i){
            cstInt curPrime=primeSeive[i];
            if(curPrime*curPrime>n1){
                Dn1*=2;
                break;
            }
            expo=1;
            while(0==(n1%curPrime)){
                ++expo;
                n1/=curPrime;
            }
            if(1<expo){
                Dn1*=expo;
            }if(1==n1){
                break;
            }
        }
        cnt=Dn*Dn1;
        Dn=Dn1;
    }
    return get_tri_num(n);
}

int main(){
    cout<<"this is the first function that isnt as fast "<<get_ans_1()<<endl;
    cout<<"this is the second function that is faster "<<get_ans_2()<<endl;
    return 0;
}

File: ProjectEuler/Python/p12.py
from math import sqrt


def gen_bit_sieve(lim):
  lim2 = lim // 2
  r = int(sqrt(lim)-1)//2
  bit_sieve = [False]*lim2
  bit_sieve[0] = True
  for i in range(1, r+1):
    if not bit_sieve[i]:
      for j in range(i*(i+1)+i*(i+1), lim2, 2*i+1):
        bit_sieve[j] = True
  return bit_sieve

def gen_sieve(lim):
  bit_sieve = gen_bit_sieve(lim)
  sieve = [2] + [i+i+1 for i in range((lim//2)-1) if not bit_sieve[i]]
  return sieve

def sol1(lim):
  t = a = 1
  cnt = 0
  tt = 0
  expo = 0
  sieve = gen_sieve(65500)
  while cnt <= lim:
    cnt = 1
    a += 1
    t += a
    tt = t
    for cur_prime in sieve:
      if cur_prime*cur_prime > tt:
        cnt += cnt
        break
      expo = 1
      while 0 == tt%cur_prime:
        expo += 1
        tt //= cur_prime
      if 1 < expo:
        cnt *= expo
      if 1==tt:
        break
  print('the value of the first triangle number to have over {} is {}'.format(lim, t))

def get_tri_num(num):
  return (num*num-num)//2

def sol2(lim):
  '''a studied solution'''
  n = 3
  dn = 2
  cnt = 0
  n1 = dn1 = expo = 0
  sieve = gen_sieve(1000)
  while cnt <= lim:
    n += 1
    n1 = n
    if 0 == n1%2:
      n1//= 2
    dn1 = 1
    for cur_prime in sieve:
      if cur_prime*cur_prime > n1:
        dn1 *= 2
        break
      expo = 1
      while 0 == n1%cur_prime:
        expo += 1
        n1 //= cur_prime
      if 1 < expo:
        dn1 *= expo
      if 1==n1:
        break
    cnt = dn*dn1
    dn = dn1
  print('the value of the first triangle number to have over {} is {}'.format(lim, get_tri_num(n))) 


def main():
  sol1(500)
  sol2(500)
main()
